### PR TITLE
Deprecate MM_ScavengerForwardedHeader over MM_ForwardedHeader

### PR DESCRIPTION
**Pull request by bragaigor**
Substitute MM_ScavengerForwardedHeader usage with
MM_ForwardedHeader since both APIs are meant to accomplish
the same goal and contain overlap functionality.

Depends: https://github.com/eclipse/omr/pull/5769

Signed-off-by: Igor Braga <higorb1@gmail.com>

### CODE REVIEW

**Comment by bragaigor** [comment]
cc: @amicic  @dmitripivkine

**Comment by gacholio** [comment]
Big thumbs up - it was unfortunate to have to go through the pain twice when I was doing the runtime compressed refs work.

File: runtime/gc_vlhgc/CopyForwardScheme.cpp [hunk 1]
```diff
@@ -646,17 +646,18 @@ MM_CopyForwardScheme::isObjectInNurseryMemory(J9Object *objectPtr)
 }
 
 MMINLINE void
-MM_CopyForwardScheme::calculateObjectDetailsForCopy(MM_ScavengerForwardedHeader* forwardedHeader, UDATA *objectCopySizeInBytes, UDATA *objectReserveSizeInBytes, bool *doesObjectNeedHash, bool *isObjectGrowingHashSlot)
+MM_CopyForwardScheme::calculateObjectDetailsForCopy(MM_ForwardedHeader* forwardedHeader, UDATA *objectCopySizeInBytes, UDATA *objectReserveSizeInBytes, bool *doesObjectNeedHash)
```
Comment (amicic) [inline runtime/gc_vlhgc/CopyForwardScheme.cpp:649]: At some point we can try to reuse GC_ObjectModelDelegate::calculateObjectDetailsForCopy, but let's keep this change to only what's must be done to deprecate SFH.

File: runtime/gc_vlhgc/CopyForwardScheme.cpp [hunk 2]
```diff
@@ -667,20 +668,18 @@ MM_CopyForwardScheme::calculateObjectDetailsForCopy(MM_ScavengerForwardedHeader*
 	}
 
 	/* IF the object has been hashed and has not been moved, then we need generate hash from the old address */
-	*doesObjectNeedHash = (forwardedHeader->hasBeenHashed() && !forwardedHeader->hasBeenMoved());
-	*isObjectGrowingHashSlot = false;
+	uintptr_t forwardedHeaderPreservedSlot = forwardedHeader->getPreservedSlot();
```
Comment (amicic) [inline runtime/gc_vlhgc/CopyForwardScheme.cpp:671]: be consistent within this large file and use UDATA

File: runtime/gc_vlhgc/CopyForwardScheme.cpp (hunk 1 above)
Comment (bragaigor) [reply runtime/gc_vlhgc/CopyForwardScheme.cpp:649]: Sounds good, and you mean keeping `isObjectGrowingHashSlot` right?

Comment (amicic) [reply runtime/gc_vlhgc/CopyForwardScheme.cpp:649]: It's ok to remove it. neither GC_ObjectModelDelegate::calculateObjectDetailsForCopy has it. No further action here. IT was mostly an FYI.

File: runtime/gc_base/GCObjectEvents.cpp [hunk 3]
```diff
@@ -31,18 +31,19 @@
 #include "mmomrhook_internal.h"
 
 #include "EnvironmentBase.hpp"
+#include "ForwardedHeader.hpp"
 #include "HeapMap.hpp"
 #include "MemorySpace.hpp"
 #include "MemorySubSpace.hpp"
 #include "MemorySubSpaceSemiSpace.hpp"
 #include "MemorySubSpaceRegionIterator.hpp"
+#include "ObjectAccessBarrier.hpp"
```
Comment (amicic) [inline runtime/gc_base/GCObjectEvents.cpp:40]: I guess, this was somehow brought from SFH, but I wonder why it was needed  - does not seem to be obvious. Perhaps it was indirectly including something else that is more specific, in which case it should be used directly instead...

Comment (bragaigor) [reply runtime/gc_base/GCObjectEvents.cpp:40]: Yes, but it seems we only need `Heap.hpp` so I'll substitute the above over `#include "Heap.hpp"`

File: runtime/gc_vlhgc/CopyForwardScheme.cpp [hunk 4]
```diff
@@ -1982,20 +1981,20 @@ MM_CopyForwardScheme::copy(MM_EnvironmentVLHGC *env, MM_AllocationContextTarok *
...
+				if(objectModel->isIndexable(destinationObjectPtr)) {
 					updateInternalLeafPointersAfterCopy((J9IndexableObject *)destinationObjectPtr, (J9IndexableObject *)forwardedHeader->getObject());
 				}
 
 				/* IF the object has been hashed and has not been moved then we must store the previous
 				 * address into the hashcode slot at hashcode offset.
 				 */
 				if (doesObjectNeedHash) {
-					UDATA hashOffset = forwardedHeader->getHashcodeOffset(_javaVM, destinationObjectPtr);
+					UDATA hashOffset = objectModel->getHashcodeOffset(destinationObjectPtr);
```
Comment (amicic) [inline runtime/gc_vlhgc/CopyForwardScheme.cpp:1994]: Also logic that eventually should be reused from ObjectModel though fixupForwardedObject(). Although we probably slightly differ how we set moved bit in Standard GCs vs Balanced GC, so we would possibly need another variant.

**Comment by amicic** [comment]
why not removing SFH files?

**Comment by bragaigor** [comment]
It seems that DDR needs them here: https://github.com/eclipse/openj9/blob/master/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V1.java#L31 and here https://github.com/eclipse/openj9/blob/master/runtime/ddr/gcddr.cpp#L75
Could I get rid of them? For instance could I substitute `FORWARDED_TAG` and `GROW_TAG` with locals constants? And delete line 75 from `gcddr.cpp`?

File: runtime/gc_base/GCObjectEvents.cpp (hunk 3 above)
Comment (amicic) [reply runtime/gc_base/GCObjectEvents.cpp:40]: Sounds good.

While trying to understand why it was needed I observed also being included in a few other files without a clear reason, including one that you touched: CopyForwardScheme.cpp. Please, check why/if it was needed there, too.

**Comment by amicic** [comment]
Ok, I see....

Well, GCScavengerForwardedHeader is already somewhat mismatching the C code. For midScavenge heap iteration case in DDR GCCheck we should be using (even though we know the name does not suggest it properly) GCForwardedHeader. But that's just naming, so we can mostly ignore it (although it would be nice to move the code to GCForwardedHeader to match C code)

The problem as you already observed is mostly about GROW_TAG. If we debug older core files, with newer DDR, we must remove the tag from forwarded header, but we do not have to do it for newer files (but it's benign if we do). 

So, yes, we could just manually define GROW_TAG or better just ALL_TAG, and always strip them. Alternatively, we could define ALL_FORWARDED_TAG in ScavengerHeader.hpp and use existing OMR_FORWARDED_TAG.

**Comment by bragaigor** [comment]
If we rename `GCScavengerForwardedHeader` everywhere with `GCForwardedHeader` DDR should still be happy right? As we're not changing any logic with that.

Regarding the `GROW_TAG`, your suggestion worked by adding `ALL_FORWARDED_TAG` in `ForwardedHeader.hpp`. In this case I'll need to open a PR in OMR project side. 

And with that, can we safely remove `MM_ScavengerForwardedHeader` from `runtime/ddr/gcddr.cpp`? Because that would be only usage left of SFH.

**Comment by amicic** [comment]
> If we rename `GCScavengerForwardedHeader` everywhere with `GCForwardedHeader` DDR should still be happy right? As we're not changing any logic with that, or would we need to leave the existing `GCScavengerForwardedHeader` files and create new ones for `GCForwardedHeader`?

I don't know how hard it's to rename it. It's only nice to do, so it if it causes any pain (you have at least 2 files to deal with), just give up.

> Regarding the `GROW_TAG`, your suggestion worked by adding `ALL_FORWARDED_TAG` in `ForwardedHeader.hpp`. In this case I'll need to open a PR in OMR project side.

Yes, we need an OMR change to introduce ALL tag, first.


> And with that, can we safely remove `MM_ScavengerForwardedHeader` from `runtime/ddr/gcddr.cpp`? Because that would be only usage left of SFH.

Hope so, do know for sure. Try before proceeding with OMR PR.

**Comment by bragaigor** [comment]
I removed all occurrences and usages of `MM_ScavengerForwardedHeader` (including the files) and all seems to be working. I'll run some more extended tests before opening the ALL tag PR in OMR side just to be double sure

**Comment by dmitripivkine** [comment]
> I removed all occurrences and usages of `MM_ScavengerForwardedHeader` (including the files) and all seems to be working. I'll run some more extended tests before opening the ALL tag PR in OMR side just to be double sure

I hope it is not applicable in this case but would you please double check? DDR required to support backwards compatibility so new DDR code should handle cores created by old code. So we must be sure we did not lose possibility to decode old cores by new DDR code.

**Comment by amicic** [comment]
OK, VM compiles and runs, and DDR compiles, but as Dmitri suggested you should test DDR a bit.... You need to create a core dump with old (before these changes) and new (with these changes, specifically, with forwarded header without GROW bit being set) and see that new DDR/jextract will be able to iterate the heap (by for example doing a GC check). 
The core file must be generated in a middle PGC.

Anyhow, since in the newest implementation of DDR, we would strip both forwarded tag (0x4) and GROW flag (0x2) tag though ALL tag (0x6), we should be ok.

Similar test you can be done with a Concurrent Scavenge (CS), what is a more complex case than Standard Scavenge, which also sets FH tags other than just FORWARDED_TAG.  it uses COPY HINT what is is same as GROW, 0x2 (that case was already covered by old DDR, by luck)

Additionally, to make DDR with CS fully happy (which is somewhat broken, currently) ALL tag should be 0x7, since CS can set OMR_SELF_FORWARDED_TAG (0x1), too. This is even harder to encounter/test. Not only core has to be generated in a middle of CS, but also that Scavenge cycle has to abort!

**Comment by bragaigor** [comment]
I’ve run several tests since yesterday and most of them passed, but DDR is not happy. In order to be able to remove `ScavengerForwardedHeader` completely the main challenge is in `GCScavengerForwardedHeader_V1.jav`a where DDR makes use of `ScavengerForwardedHeader`. I tried updating this file to use `ForwardedHeader` instead but DDR was not happy. I tested the scenario above where I generated a core file with core dump with old (before these changes) in the middle of a PGC and ran jdmpview with new version. When I ran `gccheck` with `midscavenge`, DDR was looking for `MM_ForwardedHeader.ALL_FORWARDED_TAG` (accessing existing `MM_ForwardedHeader.OMR_FORWARDED_TAG` also doesn't work) in the old DDR version which is non existent and it crashed with:
```
Checking CLASS HEAP...Error executing DDR command !gccheck all,noobjectheap:all:midscavenge,quiet : null
java.lang.reflect.InvocationTargetException
…
Caused by: java.lang.NoClassDefFoundError: com.ibm.j9ddr.vm29.j9.gc.GCScavengerForwardedHeader_V1 (initialization failure)
	at java.base/java.lang.J9VMInternals.initializationAlreadyFailed(J9VMInternals.java:135)
	at com.ibm.j9ddr.vm29.j9.gc.GCScavengerForwardedHeader.fromJ9Object(GCScavengerForwardedHeader.java:53)
…
Caused by: java.lang.NoSuchFieldError: com/ibm/j9ddr/vm29/structure/MM_ForwardedHeader.ALL_FORWARDED_TAG
	at com.ibm.j9ddr.vm29.j9.gc.GCScavengerForwardedHeader_V1.<clinit>(GCScavengerForwardedHeader_V1.java:39)
```
I then tried the following:
- We keep ScavengerForwardedHeader files and version GCScavengerForwardedHeader_V1 for .*_V2
- Remove ScavengerForwardedHeader files and update GCScavengerForwardedHeader_V1 to use local variables for ALL_TAGS and FORWARDED_TAG. And version GCScavengerForwardedHeader_V1 for .*_V2
- We could even get away without versioning and without adding new flags to `ForwardedHeader`, but I don't think that would be correct since DDR would only be aware of `ScavengerForwardedHeader` and not `ForwardedHeader` which is what we want.

Bottom line is how to access `FORWARDED_TAG` from `GCScavengerForwardedHeader_V1`. All three options worked but I would say the second one is a better option since we’re deprecating `ScavengerForwardedHeader` and we want DDR to be aware of `ForwardedHeader` instead of `ScavengerForwardedHeader`?

**Comment by bragaigor** [comment]
@keithc-ca would it be okay to go ahead with removing all usages of `ScavengerForwardedHeader` including the files and updating `GCScavengerForwardedHeader_V1` to use local variables since we're deprecating `ScavengerForwardedHeader`? I did run some backward compatibility tests and they all worked. Or would you prefer to keep those files and `GCScavengerForwardedHeader_V1` unchanged? Either way, I would be introducing a new version of `GCScavengerForwardedHeader_V2`.

**Comment by keithc-ca** [comment]
> @keithc-ca would it be okay to go ahead with removing all usages of `ScavengerForwardedHeader` including the files and updating `GCScavengerForwardedHeader_V1` to use local variables since we're deprecating `ScavengerForwardedHeader`? I did run some backward compatibility tests and they all worked. Or would you prefer to keep those files and `GCScavengerForwardedHeader_V1` unchanged? Either way, I would be introducing a new version of `GCScavengerForwardedHeader_V2`.

I think `GCScavengerForwardedHeader_V1` must remain for accessing old core files. The DDR code relies on the declaration of `MM_ScavengerForwardedHeader` so that must remain as well (it's fine if no native code uses it).

**Comment by amicic** [comment]
Well, I prefer removing native MM_ScavengerForwardedHeader altogether, and instead introducing DDR locally defined consts to have V1 compile/run. Does not make sense to drag around not so small native files for the sake of 2 const.

A compromise is to remove most of content of ScavengerForwardedHeader.hpp/cpp (probably we can completely remove *.cpp) except a couple of const definitions to make DDR happy. But again, it's cleaner to me to just define them in DDR.

**Comment by keithc-ca** [comment]
I agree it would be nice to remove the type altogether, but how do you propose that DDR will acquire the information it needs (field names, types and offsets)?

**Comment by amicic** [comment]
Right, ideally we need to dump content of ScavengerForwardedHeader in old cores, as well (I was mostly thinking about how to keep getForwardedObject working).

Ok, so this should work: we keep the ScavengerForwardedHeader.hpp and we have the class declaration with their fields and consts, but we remove all methods (inlined in hpp and remove completely cpp files)?

**Comment by bragaigor** [comment]
Did some quick backward compatibility tests and Aleks idea works. I'll run some more tests to be double sure, and once everything works I'll update this PR.

File: runtime/gc_base/ScavengerForwardedHeader.hpp [hunk 5]
```diff
@@ -62,227 +62,7 @@ class MM_ScavengerForwardedHeader
 	bool const _compressObjectReferences;
 #endif /* defined(OMR_GC_COMPRESSED_POINTERS) && defined(OMR_GC_FULL_POINTERS) */
 private:
-
```
Comment (amicic) [inline runtime/gc_base/ScavengerForwardedHeader.hpp:65]: you can remove most if not all #include

Comment (amicic) [reply runtime/gc_base/ScavengerForwardedHeader.hpp:65]: change UDATA to uintptr_t

File: runtime/gc_base/ScavengerForwardedHeader.hpp [hunk 6]
```diff
@@ -306,140 +86,7 @@ class MM_ScavengerForwardedHeader
 	{
 	}
 
-	/**
```
Comment (amicic) [inline runtime/gc_base/ScavengerForwardedHeader.hpp:309]: Did compiler forced you do have a constructor? Semantically, we (neither native nor DDR) don't need it .

File: runtime/gc_vlhgc/CopyForwardScheme.cpp [hunk 7]
```diff
@@ -667,20 +668,18 @@ MM_CopyForwardScheme::calculateObjectDetailsForCopy(MM_ScavengerForwardedHeader*
 	}
 
 	/* IF the object has been hashed and has not been moved, then we need generate hash from the old address */
-	*doesObjectNeedHash = (forwardedHeader->hasBeenHashed() && !forwardedHeader->hasBeenMoved());
-	*isObjectGrowingHashSlot = false;
+	UDATA forwardedHeaderPreservedSlot = forwardedHeader->getPreservedSlot();
```
Comment (amicic) [inline runtime/gc_vlhgc/CopyForwardScheme.cpp:671]: This assumes that preserved flags are lowest bits of preserved slot, what actually is true. Still, it's more future proof to extract the flags and then proved them later to hasBeenMoved and other APIs:

forwardedHeaderPreservedFlags = getPreservedFlags(forwardedHeader);
....

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import com.ibm.j9ddr.vm29.pointer.UDATAPointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9BuildFlags;
+import com.ibm.j9ddr.vm29.pointer.generated.J9ObjectPointer;
+import com.ibm.j9ddr.vm29.structure.J9Consts;
+import com.ibm.j9ddr.vm29.structure.MM_ForwardedHeader;
+import com.ibm.j9ddr.vm29.types.U32;
+import com.ibm.j9ddr.vm29.types.UDATA;
+import com.ibm.j9ddr.vm29.pointer.helper.J9ObjectHelper;
+
+class GCScavengerForwardedHeader_V2 extends GCScavengerForwardedHeader
```
Comment (amicic) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java:36]: could we extend this from V1, and reuse most of the methods?

Comment (bragaigor) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java:36]: For some reason I thought/hoped that in the future we would get rid of V1 but we can't do that for DDR. Will simplify this

File: runtime/ddr/gcddr.cpp [hunk 9]
```diff
@@ -72,7 +71,6 @@ GC_DdrDebugLink(MM_MemoryPoolAddressOrderedList)
 GC_DdrDebugLink(MM_MemoryPoolHybrid)
 GC_DdrDebugLink(MM_MemoryPoolSplitAddressOrderedList)
 GC_DdrDebugLink(MM_RealtimeMarkingScheme)
-GC_DdrDebugLink(MM_ScavengerForwardedHeader)
```
Comment (keithc-ca) [inline runtime/ddr/gcddr.cpp:75]: I expect this will break DDR support for older core files.

File: runtime/gc_base/ScavengerForwardedHeader.hpp [hunk 10]
```diff
@@ -23,17 +23,6 @@
 #if !defined(SCAVENGERFORWARDEDHEADER_HPP_)
 #define SCAVENGERFORWARDEDHEADER_HPP_
 
-#include "j9.h"
-#include "j9cfg.h"
```
Comment (keithc-ca) [inline runtime/gc_base/ScavengerForwardedHeader.hpp:27]: This should not assume anything is include before this file, which needs things like `uintptr_t` and `OMR_GC_COMPRESSED_POINTERS`.

**Review by amicic — APPROVED** [review]
Please squash the commits

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.ibm.j9ddr.CorruptDataException;
+import com.ibm.j9ddr.vm29.pointer.generated.J9ObjectPointer;
+import com.ibm.j9ddr.vm29.structure.MM_ForwardedHeader;
+import com.ibm.j9ddr.vm29.pointer.helper.J9ObjectHelper;
+
+class GCScavengerForwardedHeader_V2 extends GCScavengerForwardedHeader_V1
+{
+	/* combine the flags into one mask which should be stripped from the pointer in order to remove all tags */
+	private static final int ALL_TAGS = (int)MM_ForwardedHeader.OMR_FORWARDED_TAG_MASK;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java:32]: Please remove this unused field.
Also, trim any trailing whitespace from all lines.

Comment (bragaigor) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java:37]: That's my bad, I removed one of the methods that used the field. Now it's fixed.

**Comment by bragaigor** [comment]
I did a last walk-through, rebased, squashed and all major tests are passing

**Review by keithc-ca — APPROVED** [review]

**Comment by amicic** [comment]
Jenkins test sanity all jdk11

**Comment by amicic** [comment]
I was holding the merge of this change intentionally, since we had an independent problem walking heap in DDR mid-PGC, what seemed related.

Specifically, a part of this change is removal of Grow bit in FP (Forwarded Pointer in source object), seemingly useless, but we believe we now better understand its original intention. It was to recognize a special case when a hashed object was moved first time, in which case the object would grow by a hash slot. DDR when walking evacuate region needs to skip forwarded objects, and to determine size it would get info from destination object. If that size just grown this cycle, it would incorrectly skip more (by hash slot) in evacuate. If it has grown more than 1 PGC ago, both src and dst object would be same, larger size, and the skip would be correct. So, to distinguish this special case, grow bit would be set in FP, so that an iterator would know when it needs to adjust/reduce by hash slot size, relative from destination object size.

This same problem exists in Scavenger too, but it's worked around differently, using a special middle state in destination object (without going into detals how). DDR code is aware of this special state (GCScavengerForwardedHeader_V1::getObjectSize()) and should properly adjust the size to get proper src size object to skip.

DDR is however not fully aware of Grow bit (other than it just needs to strip it to get pure forwarded address) and hence, DDR is generally not capable of walking heap dumped mid PGC.

Overall, removing Grow bit is still reasonable, since it's not regressing anything. We, however, believe we can unify behavior of destination object bits for Scavenger and PGC (in a separate change) so that DDR is eventually able to walk heap mid PGC, too.
